rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,356 @@
+
+### Nonfunctional requirements
+
+Nonfunctional requirements are not directly related to the user story but may instead reflect some technical aspect of the implementation, such as latency. [Here's](https://en.wikipedia.org/wiki/Non-functional_requirement) a more comprehensive list of nonfunctional requirements.
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used authenticate the user, and interact with the DP Backend to upload files. It will also use AWS JS SDK to
+upload files from local storage to AWS. These feature will extend the existing DP Frontend.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling users requests. It will generated presigned URL and access tokens
+for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
+This additional functionality can be built into the existing DP Backend Lambda. The backend will also be responsible for
+performing quick validation of the filing being uploaded
+
+#### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in progress download jobs.
+The download jobs are used by the EC2 instances to know what files to download before uploading them to S3.
+
+##### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - a to the link to download. Use to retrieve the file from CSP.
+- collect_id - identifies the collection. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- asset_id - identifies the asset. Used to determine the storage location, and file name.
+- file size - the size of the asset to download. Used to determine the download method, and EC2 instance size.
+
+#### AWS STS
+
+The AWS STS service will be used to generate temporary credentials used to upload a file to S3. The backend
+server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
+functionality to generate a token with temporary access to upload to S3. This token is referred to as an access token.
+
+The access token generated and provided grants direct access to the S3 bucket for uploading. This will allow
+the browser APP to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used
```

Comment:
```suggestion
the browser API to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used
```
Not sure about this one..

Reply:
I'm referring to the Data Portal frontend react app that is run from the browser that interacts with our API.